api/tests/views_tests.py

```diff
@@ -201,3 +214,166 @@ def test_post_relayaddress_flagged_error(free_user, free_api_client) -> None:
+        self.path = "/api/v1/terms-accepted-user/"
+        self.fxa_verify_path = authentication.INTROSPECT_TOKEN_URL
+        self.uid = "relay-user-fxa-uid"
+
+    def _setup_client(self, token):
+        self.client = APIClient()
+        self.client.credentials(HTTP_AUTHORIZATION=f"Bearer {token}")
+
+    def tearDown(self):
+        cache.clear()
+
+    def test_no_authorization_header_returns_400(self):
+        client = APIClient()
+        response = client.post(self.path)
+
+        assert response.status_code == 400
+        assert response.json()["detail"] == "Missing Bearer header."
+
+    def test_no_token_returns_400(self):
+        client = APIClient()
+        client.credentials(HTTP_AUTHORIZATION="Bearer ")
+        response = client.post(self.path)
+
+        assert response.status_code == 400
+        assert response.json()["detail"] == "Missing FXA Token after 'Bearer'."
+
+    @responses.activate()
+    def test_invalid_bearer_token_error_from_fxa_returns_500_and_cache_returns_500(
+        self,
+    ):
+        fxa_response = _setup_fxa_response(401, {"error": "401"})
+        not_found_token = "not-found-123"
+        self._setup_client(not_found_token)
+
+        assert cache.get(get_cache_key(not_found_token)) is None
+
+        response = self.client.post(self.path)
+        assert response.status_code == 500
+        assert response.json()["detail"] == "Did not receive a 200 response from FXA."
+        assert responses.assert_call_count(self.fxa_verify_path, 1) is True
+        assert cache.get(get_cache_key(not_found_token)) == fxa_response
+
+        # now check that the code does NOT make another fxa request
+        response = self.client.post(self.path)
+        assert response.status_code == 500
+        assert response.json()["detail"] == "Did not receive a 200 response from FXA."
+        assert responses.assert_call_count(self.fxa_verify_path, 1) is True
+
+    @responses.activate()
+    def test_jsondecodeerror_returns_401_and_cache_returns_500(
+        self,
+    ):
+        _setup_fxa_response_no_json(200)
+        err_fxa_response = {"status_code": None, "json": {}}
+        invalid_token = "invalid-123"
+        cache_key = get_cache_key(invalid_token)
+        self._setup_client(invalid_token)
+
+        assert cache.get(cache_key) is None
+
+        # get fxa response with no status code for the first time
+        response = self.client.post(self.path)
+        assert response.status_code == 401
+        assert (
+            response.json()["detail"] == "Jsondecodeerror From Fxa Introspect Response"
+        )
+        assert responses.assert_call_count(self.fxa_verify_path, 1) is True
+        assert cache.get(cache_key) == err_fxa_response
+
+        # now check that the 2nd call did NOT make another fxa request
+        response = self.client.post(self.path)
+        assert response.status_code == 500
+        assert response.json()["detail"] == "Previous FXA call failed, wait to retry."
```

Comment:
question: Does this pass? The previous post got a `401` and `Jsondecodeerror` while this one gets `500` and `Previous FXA call failed` ?

Reply:
Yeah the tests passes. Here's a branch I created with before the refactor but this same error. https://github.com/mozilla/fx-private-relay/tree/pre-refactor-fxa-authentication-test